[What's New] Support dark mode for What's New announcement icons

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

The `What's New Announcements` tool supports multiple icons. Previously, we always showed a single icon for both light and dark mode on the announcements screen, which stood out in dark mode.

This PR adds support for adaptively showing either a URL-based icon, or a base64 icon, in light or dark mode according to the user's preference.

For an announcement to support this feature, it should include two `icon` definitions for each feature, with `iconType: light` and `iconType: dark` as appropriate. These icons should have the same format as one another, i.e. be either a base64 icon by uploading a file, or a url-based icon by specifying the URL.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Update the version number in the project file to 15.9
2. Build and run the app, and sign in to the app using an a11n WordPress account
3. Observe that the What's New announcement is shown, and when you switch between dark and light mode, the images change accordingly.

Note that after you dismiss the What's New announcement, you can find it again under Menu > Settings > What's New in WooCommerce, as long as you're using a version for which there is an announcement.

This test covers both the `.adaptive` and `.adaptiveRemote` cases – the test announcement is set up with the Order-related icon as a URL image, and the Payment-related icon as a base64 image.

### Further testing
Update the announcement in the tool, found in Matticspace under Tools > Developers > Mobile Announce (ensure you edit the Woo iOS announcement targeting 15.9 only). Note that the announcement will have been stored in Core Data, and while you can go back and see it using the directions above, it will not be fetched from the network again. 

To see your changes, delete the app and run it again.

#### Suggested scenarios:
I've tested all of these, and it takes a little while, so feel free to take a selection or none of them.
- No `iconType` specified with a single URL icon – uses that icon in all modes
- No `iconType` specified with a single base64 icon – uses that icon in all modes
- `iconType: light` and `iconType: dark` specified with different icon formats – uses the light icon in all modes
- URL-based icons for one or both recognised iconTypes, with unreachable URLs – uses a placeholder
- Multiple icons specified with unrecognised `iconTypes` – uses the first icon in all modes
- Only `iconType: light` specified with a single URL icon – uses that icon in all modes
- Only `iconType: light` specified with a single base64 icon – uses that icon in all modes
- Only `iconType: "some other string"` specified with a single icon – uses that icon in all modes


## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://github.com/woocommerce/woocommerce-ios/assets/2472348/1bbf6e76-95f7-40cd-be61-53d07cd1f42d


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
Add support for adaptive What’s New icons by URL

joshheald committed:
Add support for light/dark base64 icons

joshheald committed:
Fix placeholder for URL-based adaptive images

joshheald committed:
Add dark mode What’s New announcements to 15.1 release

joshheald commented:
@toupper I've put this in 15.1, but please let me know if you'd prefer that we bump it to 15.2. It's not hugely urgent, it would just be nice to have it.

joshheald committed:
Merge branch 'trunk' into task/support-dark-mode-for-whats-new-icons

toupper commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/IconListItem.swift:
```diff
@@ -10,19 +10,42 @@ struct IconListItem: View {
         /// Base64 icon
         case base64(UIImage)
 
+        case adaptiveBase64(universal: UIImage, dark: UIImage?)
```
I am wondering, is calling these parameters universal for light specific enough? After all, we could have some light mode icons that are not universal (they wouldn't look good on dark mode).

joshheald replied:
We could... but they would not be used in an inappropriate context if a dark mode image was provided. I did it this way to raise awareness that if there's no dark image, the light one would be used in that context.

I do think this is wrong though. I used "Universal" inspired by the asset catalogue, but looking back at it, that refers to the device not the colour scheme. The correct term would be "Any" based on the asset catalogue.

Perhaps the right place to change it is on the `AdaptiveAsyncImage` and `AsyncImage` structs, rather than this enum, so that future users of those are prompted that what they pass for the light image could be used in other contexts if there's an issue with the dark one.

joshheald replied:
Actually, probably both places are good.  We may start making these icons from other places, they're already used in Shipping as well as here.

I could make it `adaptiveBase64(anyAppearance: UIImage, dark: UIImage?)` if you think that's better than `universal`?

toupper replied:
> We could... but they would not be used in an inappropriate context if a dark mode image was provided. I did it this way to raise awareness that if there's no dark image, the light one would be used in that context.

That's true, we have the same in the xcassets file, Appearances are None, Any-Dark, or Any-Light-Dark. Using `adaptiveBase64(anyAppearance: UIImage, dark: UIImage?) ` as you suggested here would be an improvement indeed 👍

toupper commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/IconListItem.swift:
```diff
@@ -10,19 +10,42 @@ struct IconListItem: View {
                 KFImage(url)
                     .resizable()
+            case .adaptiveRemote(let universal, let dark):
+                AdaptiveAsyncImage(lightUrl: universal, darkUrl: dark, scale: 3) { imagePhase in
```
Is hardcoding the scale here to 3 safe enough, or we have to retrieve the scale from an environment variable?

joshheald replied:
Scale is a property of the image we get from the server, not the environment. Right now there's nothing in the What's New Announcement API response which will support a scale, though we could consider using `iconType` for this by using different keys. E.g. `light@3x` `dark@2x`. 

Given that these icons are pretty small, it seemed like a pragmatic solution to assume 3x, as that's big enough for any device.

toupper commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/AdaptiveImage.swift:
```diff
@@ -0,0 +1,17 @@
+import SwiftUI
+
+struct AdaptiveImage: View {
+    @Environment(\.colorScheme) var colorScheme
```
Do we need to add scale here in the same way we do for the `AdaptiveAsyncImage`?

joshheald replied:
Not here, because in this case we already have the `UIImage`, so the scale is already fixed. We could use the scale at the point it's created from the data (i.e. `UIImage(data: base64Data, scale: 3.0)` when we're making the image... but since it's not required for this API, assuming a 1x scale and resizing to fit its frame may be safer. WDYT?

toupper replied:
That makes sense 👍

toupper reviewed (approved):
It looks good and tests well, I added a few comments for you to consider though. 🚢 :shipit:

joshheald committed:
Rename `light` to `anyAppearance`: fallback assets

By convention from the asset library, Any Appearance denotes that the light asset could be used elsewhere if there’s an issue with loading the dark asset, or presumably, for future types of colourscheme.

Using `anyAppearance` draws a little attention to this possibility when these adaptive elements are used.
